fix: stop color mode on windows cmd/powershell

maxandersen opened the pull request:
Why:

 * by default ANSI colors are not enabled for cmd/powershell
   on Windows 10 machines so users will by default get garbled output.

This change addreses the need by:

 * Detect when running on Windows and NOT in known working windows
   emulators and there not enable colors.

Solves #1029

maxandersen committed:
fix: stop color mode on windows cmd/powershell

Why:

 * by default ANSI colors are not enabled for cmd/powershell
   on Windows 10 machines so users will by default get garbled output.

This change addreses the need by:

 * Detect when running on Windows and NOT in known working windows
   emulators and there not enable colors.

Solves #1029

maxandersen commented:
Work in progress - just opened PR to get feedback on general idea.

My windows virtual machine box been building quarkus master (without tests) for the last 30 minutes and its over midnight so i'm going to sleep while my laptop spends some cpu cycles.

Tomorrow i'll test and verify it works as expected.

dmlloyd commented:
You can create a pull request as a "draft" PR.  It's a relatively recent feature.  When you go to open the pull request, the "create pull request" button has a dropdown menu on it; choose "open as draft" from there.

gsmet commented on core/runtime/src/main/java/io/quarkus/runtime/logging/LoggingSetupRecorder.java:
```diff
@@ -35,6 +36,25 @@
+    static final boolean IS_CYGWIN = IS_WINDOWS
+            && System.getenv("PWD") != null
+            && System.getenv("PWD").startsWith("/")
+            && !"cygwin".equals(System.getenv("TERM"));
```
if this is correct, maybe it requires a comment. It feels weird to check the `TERM` is not `cygwin` for Cygwin.

maxandersen replied:
this is taken literally as is from jansi assuming they were right :)

maxandersen replied:
for what it is worth - TERM=xterm on default cygwin.

maxandersen replied:
https://github.com/fusesource/jansi/commit/bb3d538315c44f799d34fd3426f6c91c8e8dfc55 is when it was added. 

Googling around a bit seems `TERM=cygwin` used mainly to trick some other tools to think you are running cygwin when you aren't :)

maxandersen replied:
added comment and consider it correct as from jansi used by mvn and others.

gsmet commented on core/runtime/src/main/java/io/quarkus/runtime/logging/LoggingSetupRecorder.java:
```diff
@@ -73,10 +93,29 @@ public void initializeLogging(LogConfig config) {
         InitialConfigurator.DELAYED_HANDLER.setHandlers(handlers.toArray(EmbeddedConfigurator.NO_HANDLERS));
     }
 
+    private boolean workingConsole() {
```
Maybe it's not the right name? It's just about color support, not if the console is working or not.

gsmet replied:
I suggest: `hasColorSupport` or `supportsColors`

maxandersen replied:
fixed.

maxandersen committed:
fix: better explict name

maxandersen committed:
fix: add special rules for windows

Why:

 * Windows cygwin has no console thus colors are disabled without reason.

This change addreses the need by:

 * ignore System.console() test on Windows, keep it everywhere else.

maxandersen commented:
Okey, removing the WIP moniker as i finally got this tested.

Had to remove System.console() test under windows as when running on cygwin console is null.

![2019-09-18_15-58-26](https://user-images.githubusercontent.com/54129/65155224-2aebf100-da2d-11e9-8b40-7a93d8721bee.png)

screenshot above shows same app running in cygwin (left) and plain cmd.exe (right).

Optimally I would like to have a way for users to force color on in case they have set the registry flag. Suggestions welcome :)

gastaldi commented on core/runtime/src/main/java/io/quarkus/runtime/logging/LoggingSetupRecorder.java:
```diff
@@ -35,6 +36,29 @@
 @Recorder
 public class LoggingSetupRecorder {
+
+    static final boolean IS_WINDOWS = System.getProperty("os.name").toLowerCase(Locale.ENGLISH).contains("win");
```
Just for the sake of avoiding possible NPE, it would be nice to use `getProperty("os.name", "Linux")`, WDYT?

maxandersen replied:
if that happens you are on a non-java vm :)

dmlloyd commented on core/runtime/src/main/java/io/quarkus/runtime/logging/LoggingSetupRecorder.java:
```diff
@@ -35,6 +36,29 @@
+     * <a href="https://conemu.github.io">ConEmu</a> ANSI X3.64 support enabled,
+     * used by <a href="https://cmder.net/">cmder</a>
+     */
+    static final boolean IS_CON_EMU_ANSI = "ON".equals(System.getenv("ConEmuANSI"));
```
This should be `IS_WINDOWS && ...` to avoid the extra `getenv`.

maxandersen replied:
fixed in latest force push :)

dmlloyd commented on core/runtime/src/main/java/io/quarkus/runtime/logging/LoggingSetupRecorder.java:
```diff
@@ -73,10 +97,32 @@ public void initializeLogging(LogConfig config) {
+                return false;
+            } else {
+                // Must be on some Unix variant or ANSI-enabled windows terminal...
+                return true;
```
This should also return `System.console() != null` IMO.  Otherwise if you pipe through e.g. `more`, you'll get lots of garbage.

maxandersen replied:
then cygwin won't work.

maxandersen replied:
but yeah - just tested and piping goes bad...dammit.

dmlloyd replied:
It's OK if cygwin doesn't work IMO.  They can enable colors manually.

maxandersen replied:
note: `mvn compile | less` on cygwin results in color dumps too.

dmlloyd replied:
cygwin will likely soon stop returning `null` for console: https://sourceware.org/git/?p=newlib-cygwin.git;a=commit;h=169d65a5774acc76ce3f3feeedcbae7405aa9b57

maxandersen committed:
fix: give user more explicit control over color

Why:

 * Windows is a pain when it comes to guessing color capabilities.
   Its many terminals has too many variances for our guesses to be
   always right.

This change addreses the need by:

 * Make quarkus.log.console.color optional allowing users to force
   on/off and when left unset quarkus will take a best guess.

dmlloyd reviewed (approved):
Close enough, we can improve in a subsequent PR!
